org-code-javabuilder/lib/src/main/java/org/code/javabuilder/AWSContentManager.java

```diff
@@ -0,0 +1,183 @@
+    return this.projectData.getSources();
+  }
+
+  @Override
+  public String getAssetUrl(String filename) {
+    try {
+      this.loadProjectDataIfNeeded();
+    } catch (InternalServerError e) {
+      // We should only hit this exception if we try to load an asset URL before source code has
+      // been loaded, which should only be in the the case of manual testing. Log this exception but
+      // don't throw to preserve the method contract.
+      // Note / TODO: Once we fully migrate away from Dashboard sources, we can remove the
+      // loadProjectDataIfNeeded() call here and this exception handling.
+      LoggerUtils.logException(e);
+      return null;
+    }
+    return this.projectData.getAssetUrl(filename);
+  }
+
+  @Override
+  public String generateAssetUploadUrl(String filename) throws JavabuilderException {
+    if (this.uploads >= UPLOADS_PER_SESSION) {
+      throw new UserInitiatedException(UserInitiatedExceptionKey.TOO_MANY_UPLOADS);
+    }
+    final String key = this.generateKey(filename);
+    final long expirationTimeMs = System.currentTimeMillis() + context.getRemainingTimeInMillis();
+
+    try {
+      final URL presignedUrl =
+          s3Client.generatePresignedUrl(
+              this.bucketName, key, new Date(expirationTimeMs), HttpMethod.PUT);
+      this.uploads++;
+      // Add the GET url for this file to the asset map so it can be referenced later.
+      this.projectData.addNewAssetUrl(filename, this.getContentUrl + "/" + key);
+      return this.getContentUrl + presignedUrl.getFile();
+    } catch (AbortedException e) {
+      // this is most likely because the end user interrupted program execution. We can safely
+      // ignore this.
+    } catch (SdkClientException e) {
+      throw new InternalServerError(InternalErrorKey.INTERNAL_RUNTIME_EXCEPTION, e);
+    }
+
+    return null;
+  }
+
+  @Override
+  public String writeToOutputFile(String filename, byte[] inputBytes, String contentType)
+      throws JavabuilderException {
+    if (this.writes >= WRITES_PER_SESSION) {
+      throw new UserInitiatedException(UserInitiatedExceptionKey.TOO_MANY_WRITES);
+    }
+    String filePath = this.generateKey(filename);
+    ObjectMetadata metadata = new ObjectMetadata();
+    metadata.setContentType(contentType);
+    metadata.setContentLength(inputBytes.length);
+    ByteArrayInputStream inputStream = new ByteArrayInputStream(inputBytes);
+
+    try {
+      this.s3Client.putObject(this.bucketName, filePath, inputStream, metadata);
+    } catch (AbortedException e) {
+      // this is most likely because the end user interrupted program execution. We can safely
+      // ignore this.
+    } catch (SdkClientException e) {
+      // We couldn't write to S3, send a message to the user and fail. The S3 SDK includes retries.
+      throw new InternalServerError(InternalErrorKey.INTERNAL_RUNTIME_EXCEPTION, e);
+    }
+
+    this.writes++;
+    return this.getContentUrl + "/" + filePath;
```

Comment:
What is the `getContentUrl`? I was assuming it was the URL used to GET content from S3, but it looks like we're also using it for PUTs too. If that's the case, perhaps just `contentUrl` would be clearer?

Reply:
Yep good call - this is copied over from when we were only using the URL to GET theater files. I'll change to `contentUrl`

Reply:
Ahh, I'm also realizing I was reading all this in terms of getting the user sources from S3, and didn't realize these were updating existing functionality from when we added Theater and Prompter. `getContentUrl` does make a lot more sense now.